R/join-rel.R

```diff
@@ -0,0 +1,266 @@
+
+#' Append a Substrait Join relation
+#'
+#' @param compiler_left,compiler_right A [substrait_compiler()] or table-like
+#'   object to use as inputs to the join relation. Currently only one of
+#'   these can be a [substrait_compiler()].
+#' @param by A join specifiation. Join specifications that are currently
+#'   supported include a character vector of column names whose optional
+#'   names indicate the name on `compiler_left` (e.g., `"common_name"` or
+#'   `c("name_left" = "name_right")`).
+#' @param type One of JOIN_TYPE_INNER, JOIN_TYPE_OUTER, JOIN_TYPE_LEFT,
+#'   JOIN_TYPE_RIGHT, JOIN_TYPE_SEMI, JOINT_TYPE_ANTI, or JOIN_TYPE_SINGLE.
+#' @param name_repair A function of `output_mapping`, `names_left`, and
+#'   `names_right` used to calculate the output names (e.g.,
+#'   [join_name_repair_suffix_common()] or [join_name_repair_none()]).
+#' @param output_mapping A function of `by`, `names_left`, and `names_right`
+#'   used to calculate the zero-based indices of the output to include (e.g.,
+#'   [join_emit_all()] or [join_emit_default()]).
+#' @param names_left,names_right Arguments passed to the `name_repair` and
+#'   `output_mapping` functions to allow calculating reasonable output
+#'   parameters.
+#' @param suffix A length 2 character vector of suffixes to use to
+#'   disambiguate columns from the left and right inputs whose name would
+#'   be duplicated in the output.
+#'
+#' @return A modified compiler.
+#' @export
+#'
+#' @examplesIf has_duckdb_with_substrait()
+#' left <- tibble::tibble(number = 1:3, letter = c("a", "b", "c"))
+#' right <- tibble::tibble(number = 2:4, LETTER = c("B", "C", "D"))
+#'
+#' substrait_join(
+#'   duckdb_substrait_compiler(left),
+#'   right
+#' )
+#'
+substrait_join <- function(compiler_left, compiler_right, by = NULL,
+                           type = "JOIN_TYPE_INNER",
+                           name_repair = join_name_repair_suffix_common(),
+                           output_mapping = join_emit_all) {
```

Comment:
How about we rename these to `name_repair_func` and `output_mapping_func` to make it obvious that the params here should be functions?

There's something a bit unintuitive here for me, especially with the defaults having one being a called function and one being the name of a function.  I think something needs renaming or refactoring.